Update 0004-releases.md

digitalsadhu opened the pull request:
I've given the release schedule some thought after experimentation with the React repo for some time. Happy to hear peoples thoughts on what I'm proposing here.

@fabric-ds/fabric-core-team 

Rendered Version:
https://github.com/fabric-ds/issues/blob/d69448b554b93d40b824c137d391125fca02ec7a/rfcs/0004-releases.md

digitalsadhu committed:
Update 0004-releases.md

BalbinaK commented on rfcs/0004-releases.md:
```diff
@@ -57,9 +57,12 @@ The following steps would need to be taken to implement this from a technical pe
   * Release triggers a slack webhook to notify that a new release is ready with link to update guide
 
+## Schedule
+* Tradeoffs exist around releases. Too long of a period between releases can potentially cause friction for teams who need new features quickly and are forced to termporarily use the next branch. On the other hand, too short and we increase "update noise" for teams. Additionally, less frequent releases introduce more risk when updating the Eik aliases.
```
> less frequent releases introduce more risk when updating the Eik aliases.

Would patch and minor version changes introduce that risk, too?

If not, would it make sense to differentiate release schedule based on whether it's a major or minor/patch update? Just trying to think how "noisy" it could be to get notified of the latter on a weekly basis. If that was already discussed elsewhere then ignore my comment 😄

digitalsadhu replied:
There's a question, I think, around how valuable it would be to have process around major changes. The best way I've seen to handle this is to add deprecation messages during minor and patch releases and then every so often, do a major by removing deprecated features all in one go. This reduces the frequency of major releases and makes for a very clear upgrade path for users. It would require us to be pretty dedicated to the process, very thorough etc which could be challenging and introduce significant overhead so I'd be a bit reluctant.

digitalsadhu replied:
> > less frequent releases introduce more risk when updating the Eik aliases.
> 
> Would patch and minor version changes introduce that risk, too?

Any changes introduce the risk but small and often is always easier to handle than larger and less often.

pearofducks commented:
I think I'm missing the core "why" this is beneficial. The background section says,

> new versions of our design system mean that teams who are busy working on functionality for Finn need to take time out of their schedules to upgrade

but is that true for the vast majority of our releases? Between Renovate and Eik, aren't teams getting the latest patch/minor versions for free the second we would push to main?

---

Doing this for majors at least makes some sense to me. Since _those_ benefit from having some scheduling/reliability/predictability and ofc imply API changes, thus teams need to put in work to get current.

Looking at this with the lone-frontend-developer-on-my-team hat on 🎩 , I would probably ignore all the release notes until there was some action required of me. So this actually worsens the signal/noise ratio maybe?

pearofducks commented:
I would love to see this specific scenario addressed as well, are there maybe "cons" to this approach that need "mitigations"?

(Again with my team-developer hat on) If you told me a non-breaking change was on `next` that I needed to get an app out, and I needed to switch my app to use that tag, then remember at some point in the future I needed to switch back; I would think you were a bit insane and probably argue with you. Why are we putting this thing that doesn't break other people's apps on another tag, and making devs who need that thing right away go through extra effort to use it?

digitalsadhu commented:
> I think I'm missing the core "why" this is beneficial. The background section says,
> 
> > new versions of our design system mean that teams who are busy working on functionality for Finn need to take time out of their schedules to upgrade
> 
> but is that true for the vast majority of our releases? Between Renovate and Eik, aren't teams getting the latest patch/minor versions for free the second we would push to main?
> 
> Doing this for majors at least makes some sense to me. Since _those_ benefit from having some scheduling/reliability/predictability and ofc imply API changes, thus teams need to put in work to get current.
> 
> Looking at this with the lone-frontend-developer-on-my-team hat on 🎩 , I would probably ignore all the release notes until there was some action required of me. So this actually worsens the signal/noise ratio maybe?

I think the main idea I want to see here is that we try to reduce the amount developers even need to notice what's going on day to day or even week to week. Much better that once a month there's a clear single list of features and fixes in front of them than random through the days of the weeks, mixed in with renovate bot dependency updates. I believe it would be much better for someone to see a single update document less frequently and be able to look through it and make decisions about what if anything they need to do than to get in the habit of deleting a series of email updates from renovate bot regarding Fabric. I think that the more we can bundle things into a larger release, with good release documentation etc, the more useful this will be for developers.

digitalsadhu commented:
> I would love to see this specific scenario addressed as well, are there maybe "cons" to this approach that need "mitigations"?
> 
> (Again with my team-developer hat on) If you told me a non-breaking change was on `next` that I needed to get an app out, and I needed to switch my app to use that tag, then remember at some point in the future I needed to switch back; I would think you were a bit insane and probably argue with you. Why are we putting this thing that doesn't break other people's apps on another tag, and making devs who need that thing right away go through extra effort to use it?

I think this switching branches problem is the one fatal floor in the plan. I'd almost rather we lengthened time between offical releases (merges from next to main) and loosened things up a bit WRT to getting urgent features out. There's already a section in this document for agreeing to get urgent bug/security fixes out. Perhaps extending this to include urgent features and doing away with the idea of asking people to switch channels would be the best approach.

kjesle commented on rfcs/0004-releases.md:
```diff
@@ -57,9 +57,12 @@ The following steps would need to be taken to implement this from a technical pe
 
+## Schedule
+* Tradeoffs exist around releases. Too long of a period between releases can potentially cause friction for teams who need new features quickly and are forced to termporarily use the next branch. On the other hand, too short and we increase "update noise" for teams. Additionally, less frequent releases introduce more risk when updating the Eik aliases.
+* A weekly schedule allows for a quick enough turn around such that most teams would not need to use the next branch (up to a week wait at most) while preventing ongoing development noise and allowing for clean communication of new releases. Eg. "New releases of Fabric every Monday at 10am"
```
I'm not really sure I get why a schedule is needed. For bug-fixes and the like, it should just be released into master and released automatically at once.

For major changes that should be released when it is ready and not "rushed to Monday or we lose the window."

digitalsadhu replied:
I think that's fair. If we simply work against the next branch (importantly including renovate PRs there) and then just merge to main whenever necessary (which would mean whenever there's a bug fix or new feature) then I think we'd be 99% there. 

What I haven't made clear in the doc here is that we also need to update the Eik alias for each library as well and that we would like to add a layer of safety into this in the form of a delay (maybe a week). This is because when we update the alias for a library, everyone will be updated to the latest version of the library in production without the need to redeploy (and therefore rerun CI). We can roll this back in the case of issues but currently there's a delay as caches expire and such so its not as quick a rollback as in pipeline. We never bump aliases across major version boundaries so in theory the risk is low, but immediately rolling out alias changes whenever we publish a new version of say Fabric React risks that someone make a mistake and some bad code snuck out before we could catch it. Our initial idea was that a delay of one week from release would give plenty of time to detect any issues and fix before the alias was updated. A fixed release schedule makes this lagging alias update process easier to manage by avoiding the need for book keeping around how long a week is from the last release but that's not necessarily a show stopper.

vsandvold commented:
From what I understand, the proposal is to merge changes from the `next` branch into `main` on a weekly basis. These changes will then be published as new versions of `@fabric-ds` packages on NPM. In the case of major version changes, the Eik aliases will also be updated, eg. a new version number to prevent breaking changes to propagate to all apps that use Eik asset mapping. This includes asset maps for both Fabric CSS and Fabric component libs (react, vue, elements).

For apps that use Eik, I see no immediate consequences with weekly updates. Non-breaking changes will propagate automatically like before, and breaking changes will not take effect before the import map is changed in the app `eik.json`.

I'm not sure if a weekly schedule is more important for apps that use versioned Fabric packages. I tend to ignore PR's for automatic dependency updates in the repos I'm responsible for. This is mainly because I don't trust our tests to catch stuff breaking between version updates, and because of some old, bad habits. I tend to bump dependencies once in a while, which is barely manageable because the number of repos is fairly small. Bad habits die hard 🙈

digitalsadhu commented:
> In which order do we do the release? Do we write an upgrade guide before we merge the PR from next into main/master, and update the release notes with the link to the upgrade guide?

New features will be documented in the documentation and I'm unsure whether we need to actually include any update information about this as part of the release. Fixes shouldn't require anything of the user. This leaves breaking changes which we aim to do as little as humanly possible. We might therefore be able to keen this part of the process lite and only include upgrade information with breaking changes and just inline them in the GitHub release notes. Thoughts?

> I also think we should consider adding git hooks, so that one is not allowed to commit unless a few checks have passed. E.g. prettier and lint must pass, and the check ensures you are following the correct commit message format. Commitizen is nice, but it doesn't enforce anything and is merely just a support tool.

Yup, makes sense. FWIW I would prefer to catch this stuff on the CI than locally (using lint staged for example)

digitalsadhu commented:
> From what I understand, the proposal is to merge changes from the `next` branch into `main` on a weekly basis. These changes will then be published as new versions of `@fabric-ds` packages on NPM. In the case of major version changes, the Eik aliases will also be updated, eg. a new version number to prevent breaking changes to propagate to all apps that use Eik asset mapping. This includes asset maps for both Fabric CSS and Fabric component libs (react, vue, elements).

Yes, though everything in the proposal is up for feedback and change. We've been in a long experimental phase on this and any thoughts around the value of this kind of process is welcome. 

> For apps that use Eik, I see no immediate consequences with weekly updates. Non-breaking changes will propagate automatically like before, and breaking changes will not take effect before the import map is changed in the app `eik.json`.

Yes, new aliases for major versions can be created right away. For patch and minor versions, we think its worth taking extra care and leaving it a x amount of time between the release and the alias update (a week most likely) to allow for testing changes in case of a breaking change sneaking its way into a patch or minor when it shouldn't have. This might be less needed once we can do instant Eik alias rollback (currently cache timeouts mean rollbacks can take up to 30 mins.

> I'm not sure if a weekly schedule is more important for apps that use versioned Fabric packages. I tend to ignore PR's for automatic dependency updates in the repos I'm responsible for. This is mainly because I don't trust our tests to catch stuff breaking between version updates, and because of some old, bad habits. I tend to bump dependencies once in a while, which is barely manageable because the number of repos is fairly small. Bad habits die hard 🙈

One of the benefits of the process defined in this document (with or without scheduled release timing) is that we gather up all the renovate bot noise on the next branch before releasing it all in one go to main. This avoids the constant releases generated by renovate PR merges.

pearofducks commented:
> Yup, makes sense. FWIW I would prefer to catch this stuff on the CI than locally (using lint staged for example)

Agree, please no local git hooks!

> For patch and minor versions, we think its worth taking extra care and leaving it a x amount of time between the release and the alias update (a week most likely) to allow for testing changes in case of a breaking change sneaking its way into a patch or minor when it shouldn't have.

What does this `x amount of time` look like? Is this time provided by the gap between a `next` merge and when it hits main? If so, are we really getting that benefit at all if no one uses that `next` release (or maybe one app does)?

pearofducks commented:
> One of the benefits of the process defined in this document (with or without scheduled release timing) is that we gather up all the renovate bot noise on the next branch before releasing it all in one go to main.

Is there a way we can get this benefit without all the 'next' and 'main' stuff? I think this is a nice benefit, and the next/main thing is a massive drawback for most of our changes.

digitalsadhu commented:
> What does this `x amount of time` look like? Is this time provided by the gap between a `next` merge and when it hits main? If so, are we really getting that benefit at all if no one uses that `next` release (or maybe one app does)?

No, its supposed to be time on main/master. A week should mean that many people will get renovate PRs that update their version and that at least some of them test their apps in that time and report any issues. If its still radio silence after a week then it should be fine to update the alias. If this every proved not to be enough (ie we had multiple incidents where issues slipped through) then we could consider some form of smoke tests of our own but I doubt we'd need to.

digitalsadhu commented:
> > One of the benefits of the process defined in this document (with or without scheduled release timing) is that we gather up all the renovate bot noise on the next branch before releasing it all in one go to main.
> 
> Is there a way we can get this benefit without all the 'next' and 'main' stuff? I think this is a nice benefit, and the next/main thing is a massive drawback for most of our changes.

What do you feel is most painful about the next/main thing? If we were to drop any form of scheduled releasing and just merged next to main whenever we had new features/fixes to release, is that still painful?

pearofducks commented:
> What do you feel is most painful about the next/main thing? 

Mostly what I've mentioned before; the overhead this creates with messaging/asking users to switch branches/etc.

> If we were to drop any form of scheduled releasing and just merged next to main whenever we had new features/fixes to release, is that still painful?

Nope! I think if users don't need to relate to that at all and can treat Fabric packages more or less like any 'normal' NPM package that'd be awesome.

digitalsadhu commented:
> > What do you feel is most painful about the next/main thing?
> 
> Mostly what I've mentioned before; the overhead this creates with messaging/asking users to switch branches/etc.
> 
> > If we were to drop any form of scheduled releasing and just merged next to main whenever we had new features/fixes to release, is that still painful?
> 
> Nope! I think if users don't need to relate to that at all and can treat Fabric packages more or less like any 'normal' NPM package that'd be awesome.

Hmm, slightly crossed wires here perhaps, I meant to imply that by dropping scheduling, that would also drop the need for users to ever switch to next. Without scheduling, ie. releasing as needed, we could merge next to main to cut a release when features and fixes are available. We still get the advantages of grouping as we need as well. That is, renovate PRs don't get spewed out onto the main channel as they come in and we could curate several fixes or feature PRs together into a release if we felt that were appropriate. Additionally, by getting the next release, we get a proper package that we ourselves can test thoroughly before we release to a wider audience. Also, If nothings going on from our end, nothing gets released (including renovate noise) until we see fit etc so everything stays quiet on the release front.

digitalsadhu commented:
> Nope! I think if users don't need to relate to that at all and can treat Fabric packages more or less like any 'normal' NPM package that'd be awesome.

Hah! Actually its me who's wires are crossed! Misread you there. Looks like you were agreeing with me.. I think 😆

martinetruls commented:
I'm struggeling to understand how the releases/next-release works with the eik aliases. If I would be needing a next-release, what kind of work would I have to do related to import mapping? And for major and minor releases, will each team actually have to update their package, or will it kind of automatically be included because of the eik alias?

digitalsadhu commented:
> I'm struggeling to understand how the releases/next-release works with the eik aliases. If I would be needing a next-release, what kind of work would I have to do related to import mapping? And for major and minor releases, will each team actually have to update their package, or will it kind of automatically be included because of the eik alias?

For minor and patch releases, initially, the idea was that if you desperately needed a fix for something that was only available on next, you would switch to a next release by running `npm install @fabric-ds/react@next` until the fix or feature was fully released on main. While in this state, you would be bypassing the aliasing and once you moved back to main, you would automatically go back to using aliasing again. It's been pointed out though that having teams need to switch channels is not ideal so given the discussion so far above we are leaning toward never expecting users to have to change channels. In either case, users shouldn't need to do anything additional for import mapping.

For major releases, users will always need to update the alias url in eik.json manually. We will never automatically update across major version boundaries.

digitalsadhu committed:
Reword to reflect RFC discussion

Updates the content of the RFC doc to reflect the various discussions we've had regarding how we do releases. We've dropped the idea of a fixed release schedule, added some information about Eik aliasing etc.

digitalsadhu commented:
I've made an update to the RFC text based on all the great discussions here. I'll give it a day or 2 for people to review and further comment if necessary. If there's nothing further after that, I'll merge.

pearofducks commented:
[Link to a rendered diff'd version](https://github.com/fabric-ds/issues/pull/102/commits/97c336034197a0206c55569203e9b4f51c79a3a5?short_path=825878b#diff-825878b650587d26ec98e49e190459c1e6301407daed9fd173abaf5b99dae456)

LGTM!
